src/lua/uri.c

```diff
@@ -0,0 +1,336 @@
+	 * one URI in it.
+	 */
+	if (uri_count <= 1) {
+		rc = luaT_uri_create(L, &uri, idx);
+		if (rc == 0)
+			uri_set_add(uri_set, &uri);
+		return rc;
+	}
+	/*
+	 * All numeric keys corresponds to URIs in string or table
+	 * format.
+	 */
+	for (int i = 0; i < size && rc == 0; i++) {
+		lua_rawgeti(L, idx, i + 1);
+		rc = luaT_uri_create(L, &uri, lua_gettop(L));
+		if (rc == 0)
+			uri_set_add(uri_set, &uri);
+		lua_pop(L, 1);
+	}
+	if (rc != 0)
+		return rc;
+
+	/*
+	 * Here we are only in case when uri_count greater than one,
+	 * so is shouldn't be "uri" and "params" field here.
+	 */
+	if (is_field_present(L, "uri", idx)) {
+		diag_set(IllegalParams, "Invalid URI table: "
+			 "expected {uri = string, params = table}, "
+			 "{string, params = table} or "
+			 "string, {uri, params = table}");
+		return -1;
+	}
+
+	if (is_field_present(L, "params", idx)) {
+		diag_set(IllegalParams, "URI query parameters are "
+			 "not allowed for multiple URIs");
+		return -1;
+	}
+
+	for (int i = 0; i < uri_set->uri_count && rc == 0; i++) {
+		struct uri *uri = &uri_set->uris[i];
+		lua_pushstring(L, "default_params");
+		lua_rawget(L, idx);
+		rc = uri_add_params_from_lua(uri, L, false);
+		lua_pop(L, 1);
+		assert(rc == 0 || !diag_is_empty(diag_get()));
+	}
+	return rc;
+}
+
+/**
+ * Create @a uri_set from the value at the given valid @a idx, which should be
+ * positive value.
+ */
+static int
+luaT_uri_set_create(struct lua_State *L, struct uri_set *uri_set, int idx)
+{
+	int rc = 0;
+	uri_set_create(uri_set, NULL);
+	if (lua_isstring(L, idx)) {
+		rc = uri_set_create(uri_set, lua_tostring(L, idx));
+		if (rc != 0) {
+			diag_set(IllegalParams, "Incorrect URI: "
+				 "expected host:service or "
+				 "/unix.socket");
+		}
+	} else if (lua_istable(L, idx)) {
+		rc = uri_set_create_from_lua_table(uri_set, L, idx);
+	} else if (!lua_isnil(L, idx)) {
+		diag_set(IllegalParams, "Incorrect type for URI: "
+			 "should be string, number or table");
+		rc = -1;
+	}
+	assert(rc == 0 || !diag_is_empty(diag_get()));
+	return rc;
+}
+
+static int
+lbox_uri_create(lua_State *L)
+{
+	int rc = -1;
+	struct uri *uri = (struct uri *)lua_topointer(L, 1);
+	if (uri != NULL) {
+		rc = luaT_uri_create(L, uri, 2);
+		/**
+		 * We don't call lua_error to maintain
+		 * backward compatibility.
+		 */
+		if (rc != 0)
+			diag_clear(diag_get());
+	}
+	lua_pushnumber(L, rc);
+	return 1;
+}
+
+static int
+lbox_uri_set_create(lua_State *L)
+{
+	int rc = -1;
+	struct uri_set *uri_set = (struct uri_set *)lua_topointer(L, 1);
+	if (uri_set != NULL) {
+		rc = luaT_uri_set_create(L, uri_set, 2);
+		/**
+		 * We don't call lua_error to maintain
+		 * backward compatibility.
+		 */
+		if (rc != 0)
+			diag_clear(diag_get());
+	}
+	lua_pushnumber(L, rc);
+	return 1;
+}
+
+void
+tarantool_lua_uri_init(struct lua_State *L)
+{
+	static const struct luaL_Reg uri_methods[] = {
+		{NULL, NULL}
+	};
+	luaL_register_module(L, "uri", uri_methods);
```

Comment:
Nit: wouldn't this work?

```lua
lua_getfield(L, LUA_GLOBALSINDEX, "uri");
```

Reply:
No see Lua_register module we need to do something like this
```lua
lua_getfield(L, LUA_REGISTRYINDEX, "_LOADED");
lua_newtable(L);
lua_pushvalue(L, -1);
lua_setfield(L, -3, modname);  /* _LOADED[modname] = new table */
```
and add some checks that there is no module with same name. I think it's better to use already realized function

Reply:
OK, let's leave it as is.